Smt/Tactic/EqnDef.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2022 by the authors listed in the file AUTHORS and their
+institutional affiliations. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wojciech Nawrocki
+-/
+import Lean.Meta.Basic
+import Lean.Elab.Tactic.Basic
+import Lean.Elab.Term
+
+/-! Utilities for handling "equational definitions". An equational definition is an equation
+of the form `∀ x₁ ⋯ xₙ, c x₁ ⋯ xₙ = body[x₁,⋯,xₙ]` in the local context, where `c` is either
+a global constant or a local variable. The LHS should be fully applied so that the equality
+is not at a universally quantified (aka function) type.
+
+These usually start off as copied equational theorems of global constants but are then transformed
+to an SMT-compatible form.
+
+Note that equational definitions have an advantage over let-bindings in that we do not need
+to justify the termination of `body`. This does however imply that rewriting by an equational
+definition may never reach a fixpoint.
+
+## Representation
+
+The equational definition for `foo` is stored as `foo.def` in the local context.
+
+TODO(WN): We could have a custom `smt` tactic mode which provides explicit storage for eqn-defs
+and displays them more nicely. -/
+
+namespace Smt
+open Lean Meta Elab Term Tactic
+
+/-- The user name of an equational definition for 'nm'. -/
+def eqnDefName (nm : Name) : Name :=
+  nm ++ `def
+
+def throwNotEqnDef [Monad m] [MonadError m] (e : Expr) : m α :=
+  throwError "expected equational definition, got{indentD e}"
+
+/-- Get the head symbol `c` of an equational definition. -/
+def getEqnDefHead (eqn : Expr) : MetaM Expr :=
+  forallTelescopeReducing eqn fun _ eqn => do
+    let some (_, lhs, _) := eqn.eq? | throwNotEqnDef eqn
+    return lhs.getAppFn
+
+/-- Get the body of an equational definition as a lambda `fun x₁ ⋯ xₙ => body[x₁,⋯,xₙ]`. -/
+def getEqnDefLam (eqn : Expr) : MetaM Expr :=
+  forallTelescopeReducing eqn fun args eqn => do
+    let some (_, _, body) := eqn.eq? | throwNotEqnDef eqn
+    mkLambdaFVars args body
+
+/-- If `nm` has an equational definition, `getEqnDefLam` it. -/
+def getEqnDefLamFor? (nm : Name) : MetaM (Option Expr) := do
+  let some ld := (← getLCtx).findFromUserName? (eqnDefName nm) | return none
+  getEqnDefLam ld.type
+
+/-- Add an equational definition `$nm.def : ∀ x₁ ⋯ xₙ, $nm x₁ ⋯ xₙ = body[x₁,⋯,xₙ]`
+for a constant and return the equation's fvar. -/
+def addEqnDefForConst (nm : Name) : TacticM FVarId := do
+  let some eqnThm ← getUnfoldEqnFor? (nonRec := true) nm
+    | throwError "failed to retrieve equation theorem for '{nm}'"
+  let eqnInfo ← getConstInfo eqnThm
+  let (eqn, pf) ← forallTelescopeReducing eqnInfo.type fun args eqn => do
+    let some (_, lhs, body) := eqn.eq? | throwNotEqnDef eqn
+    let pf ← mkAppOptM eqnInfo.name (args.map some)
+    -- In case the definition is curried, apply it fully.
+    forallTelescopeReducing (← inferType lhs) fun args' _ => do
```

Comment:
Could you please explain what's going on in line 67? Is applying `forallTelescopeReducing` once not enough? If so, why is applying it twice enough?